### DISCUSSION TITLE
Pino logging with Express server

### DISCUSSION THREAD

**Discussion by caprica** [remix-run/remix#9722] [Q&A] [answered]
I am using Remix with the official Express stack.

I am using Pino logging for the application logging, this is all set up and working fine with JSON formatted output, eventually to be consumed by some cloud log service.

What I am looking for now is any straightforward may to switch the Express request and error logging from the default console output to either a separate Pino log destination, or to have it merged with my application log messages - but in this case to have it also format the Express logs as JSON.

Is anyone doing this? Is it even good practice to do this?

This is similar to #2935, but that discussion is four years old, so I'm looking for any contemporary best practice here.

**Comment by kiliman** [comment]
To update request logging, just replace the call to `app.use(morgan("tiny"));` with Pino.

And for error logging, you need to add a `handleError` export to your _entry.server.tsx_ file.

```ts
export function handleError(
  error: unknown,
  {
    request,
    params,
    context,
  }: LoaderFunctionArgs | ActionFunctionArgs
) {
  if (!request.signal.aborted) {
    // log to pino here
  }
}
```
https://remix.run/docs/en/main/file-conventions/entry.server#handleerror

**Reply by caprica** [reply, marked as answer]
In case anyone else needs the detail of what to do to make this work:

```bash
npm i pino-http
```

In the existing `server.ts` file that was generated by the Express template:

```javascript
import logger from 'pino-http'
// import morgan from 'morgan'
```

Later in the same `server.ts` file:

```javascript
app.use(logger()) // you can pass logger options here to configure the logger, defaults are used here
// app.use(morgan('tiny'))
```

By default it will log everything, this is a slightly abridged example showing what you can expect to see when it's working:

```
20:32:56.788] INFO: request completed
    req: {
      "id": 2,
      "method": "GET",
      "url": "/",
      "query": {},
      "params": {},
      "headers": {
        "host": "localhost:3000",
        "connection": "keep-alive",
        "cache-control": "max-age=0",
        "dnt": "1",
        "upgrade-insecure-requests": "1",
        "user-agent": "Mozilla/5.0",
        "accept": "text/html,application/xhtml+xml,application/xml",
      },
      "remoteAddress": "::1",
      "remotePort": 60880
    }
    res: {
      "statusCode": 200,
      "headers": {
        "access-control-allow-origin": "*",
        "content-type": "text/html; charset=utf-8",
        "vary": "Accept-Encoding",
        "content-encoding": "gzip"
      }
    }
    responseTime: 91
```

You can customise the output in the usual way with Pino, using a custom formatter or whatever by passing options when the logger is created.

And then as previously stated by kiliman make sure to also use a Pino logger in the error handler in your `entry.server.tsx` file.
